### DISCUSSION TITLE
Request for further testing of remapping images in non-SRGB colorspaces

### DISCUSSION THREAD

**Discussion by P-Star7** [ImageMagick/ImageMagick#5900] [Help]
### ImageMagick version

7.1.0-55

### Operating system, version and so on

Windows 10 64-bit

### Description

I recently requested this feature and have updated ImageMagick accordingly:
[This](https://github.com/ImageMagick/ImageMagick/commit/6ea94f252d4a03e24eb3ef31cdada95667de3aca
)
I have not had any luck in producing images that look any different when remapped when in jzazbz colorspace - either by turning the input and palette files into jzazbz-colorspace .tiff files (input.png -set colorspace jzazbz input.tiff) and then remapping (magick input.tiff -remap palette.tiff output.tiff), or by using SRGB .pngs and setting the -colorspace jzazbz modifier before remapping (magick input.png -colorspace jzazbz -remap palette.png -colorspace srgb output.png).

Can someone else please test this to see if they can produce images that look differently when quantized in jzazbz space vs. SRGB space, and if so, what command they used? I do not want to file a bug report quite yet as it may be that I am not using the correct commands or images that would have a difference when remapped in the two colorspaces.

**Comment by fmw42** [comment]
Post a good high quality input and your output.

**Comment by snibgo** [comment]
As fmw42 says, please show or link to your files.

I wouldn't expect either of the OP commands to work. When writing JzAzBz data to PNG, IM will first convert it to sRGB. The data can be written without conversion to TIFF, but the fact that the encoding is JzAzBz is not written to the TIFF file.

**Reply by P-Star7** [reply]
![portraits](https://user-images.githubusercontent.com/26371167/209255822-807796fa-469f-4646-9d23-6c981f4d1386.png)
![rockmanpalette](https://user-images.githubusercontent.com/26371167/209255823-439629fb-5615-4622-8f91-a9c716f58a1a.png)
Input, palette image to remap the former to. Sorry if low-res pixel art doesn't count as high-quality, but that's my use case.

> I wouldn't expect either of the OP commands to work. When writing JzAzBz data to PNG, IM will first convert it to sRGB. The data can be written without conversion to TIFF, but the fact that the encoding is JzAzBz is not written to the TIFF file.

They work in that they produce portraits.png remapped to rockmanpalette.png - but setting -colorspace before -remap does not seem to affect the output file any. It produces identical remapping outputs when -colorspace is set to sRGB or JzAzBz. I thought that's what the pull request I linked was supposed to change - to have -remap be sensitive to the current -colorspace.

**Comment by snibgo** [comment]
From the OP, the commands tried for remapping were:
```
magick input.tiff -remap palette.tiff output.tiff

magick input.png -colorspace jzazbz -remap palette.png -colorspace srgb output.png
```
Neither TIFF nor PNG can have a colorspace of JzAzBz.

Suppose you have an input imageA.ext that is sRGB, and you want to remap it using colours in imageB.ext that is also sRGB, but you want the remapping to occur in JzAzBz. I suggest you convert both images to JzAzBz, do the remapping, and convert back to sRGB. Don't save intermediate results to files. Instead, do all the work in one single IM command. Like this, Windows BAT format, untested:
```
magick ^
  ( imageA.ext -colorspace JzAzBz -set colorspace sRGB ) ^
  ( imageB.ext -colorspace JzAzBz -set colorspace sRGB -write mpr:MAP +delete ) ^
  -remap mpr:MAP ^
  -set colorspace JzAzBz -colorspace sRGB ^
  out.ext
```

**Reply by P-Star7** [reply]
> From the OP, the commands tried for remapping were:
> 
> ```
> magick input.tiff -remap palette.tiff output.tiff
> 
> magick input.png -colorspace jzazbz -remap palette.png -colorspace srgb output.png
> ```
> 
> Neither TIFF nor PNG can have a colorspace of JzAzBz.
> 
> Suppose you have an input imageA.ext that is sRGB, and you want to remap it using colours in imageB.ext that is also sRGB, but you want the remapping to occur in JzAzBz. I suggest you convert both images to JzAzBz, do the remapping, and convert back to sRGB. Don't save intermediate results to files. Instead, do all the work in one single IM command. Like this, Windows BAT format, untested:
> 
> ```
> magick ^
>   ( imageA.ext -colorspace JzAzBz -set colorspace sRGB ) ^
>   ( imageB.ext -colorspace JzAzBz -set colorspace sRGB -write mpr:MAP +delete ) ^
>   -remap mpr:MAP ^
>   -set colorspace JzAzBz -colorspace sRGB ^
>   out.ext
> ```

![out](https://user-images.githubusercontent.com/26371167/209268157-40fbecce-778a-4ded-b775-d0f26f5ce267.png)
Still not right.
Also, here is a bigger version of the input palette image to replicate this, if you are having trouble downloading the smaller one.
![rockmanpalette](https://user-images.githubusercontent.com/26371167/209268448-2a009189-9898-4ee2-8c91-ae1f661fc279.png)

**Comment by snibgo** [comment]
> Still not right.

Please be specific. What is not right?

When I try these two commands, I get different outputs:
```
%IMG7%magick ^
  imageAforremap.png ^
  -remap imageBforremap.png ^
  outA.png

%IMG7%magick ^
  ( imageAforremap.png -colorspace JzAzBz -set colorspace sRGB ) ^
  ( imageBforremap.png -colorspace JzAzBz -set colorspace sRGB -write mpr:MAP +delete ) ^
  -remap mpr:MAP ^
  -set colorspace JzAzBz -colorspace sRGB ^
  outB.png
```
The outputs are:

Tested on IM version 7.1.0-49.
![outA](https://user-images.githubusercontent.com/33812368/209269263-f0d43fb8-bebc-444a-badd-3067bc4c06f0.png)
![outB](https://user-images.githubusercontent.com/33812368/209269283-8bb76a15-aad8-4f14-90fc-00401176de25.png)

**Reply by P-Star7** [reply]
Sorry, what I mean by "not right" is, if you remapped it using SRGB (magick portraits.png -remap rockmanpalette.png out.png), you get this image that is much closer to the original image.
![out](https://user-images.githubusercontent.com/26371167/209269692-42d2b60b-c514-420f-ae81-a1c132052150.png)
The colours have been reduced to that of rockmanpalette.png but without ugly black blotches. I can't fathom why it would be different between the colour spaces. The colorspace operator should just determine how the remapping is DONE, right? So it should still be mapping RGB-gamut colours to RGB-gamut colours, just in a better way due to JzAzBz's better perceptual distance. So there SHOULD be no "out of gamut" colours, right?

**Comment by snibgo** [comment]
I think I used the wrong input image for my tests. Using the first image you showed in this thread, my two outputs are:
![outA](https://user-images.githubusercontent.com/33812368/209274328-60323b7c-1c9f-4a3b-a320-7cda818eb3b1.png)

![outB](https://user-images.githubusercontent.com/33812368/209274354-226d1529-6c4b-4443-ae63-b9723df6586c.png)

As you say, doing the remapping in JzAzBz maps more pixels to black. It is supposedly more perceptually uniform than CIELab, but is also massively larger. This means that sRGB colours map to a very small range of values in JzAzBz. This might cause problems when remapping.

> The colorspace operator should just determine how the remapping is DONE, right?

No. "-colourspace XXX" changes pixels values to XXX colourspace (where "XXX" is JzAzBz or Lab or whatever), and it also sets the metadata. In my sample command, I also use "-set colorspace XXX" which changes the metadata without also changing pixel values. In my commands, I do that to prevent "-remap" from changing color spaces.

**Reply by P-Star7** [reply]
So why did [this](https://github.com/ImageMagick/ImageMagick/issues/5731#issuecomment-1306057660) work? It seems as though mapping in JzAzBz did give better results than RGB for that one.
